src/games/strategy/triplea/util/TransportUtils.java

```diff
@@ -223,4 +243,74 @@ private static void loadMaxUnits(final Unit transport, final List<Unit> canBeTra
     }
   }
 
+  private static Map<Unit, List<Unit>> findTransportsThatUnitsCouldUnloadFrom(final Collection<Unit> units,
+      final Collection<Unit> transports) {
+    final List<Unit> canBeTransported = Match.getMatches(units, Matches.UnitCanBeTransported);
+    final List<Unit> canTransport = Match.getMatches(transports, Matches.UnitCanTransport);
+    final Map<Unit, List<Unit>> result = new LinkedHashMap<>();
+    for (final Unit unit : canBeTransported) {
+      final List<Unit> transportOptions = new ArrayList<>();
+      for (final Unit transport : canTransport) {
+        if (containsEquivalentUnit(unit, TransportTracker.transporting(transport))) {
+          transportOptions.add(transport);
+        }
+      }
+      result.put(unit, transportOptions);
+    }
+    return result;
+  }
+
+  private static Map<Unit, List<Unit>> sortByTransportOptionsAscending(
+      final Map<Unit, List<Unit>> unitToPotentialTransports) {
+    final Map<Unit, List<Unit>> result = new LinkedHashMap<>();
+    unitToPotentialTransports.entrySet().stream()
+        .sorted((o1, o2) -> o1.getValue().size() - o2.getValue().size())
+        .forEachOrdered(e -> result.put(e.getKey(), e.getValue()));
+    return result;
+  }
+
+  private static Unit findOptimalTransportToUnloadFrom(final Unit unit,
+      final Map<Unit, List<Unit>> unitToPotentialTransports) {
+    double minAverageTransportOptions = Integer.MAX_VALUE;
+    Unit selectedTransport = unitToPotentialTransports.get(unit).get(0);
+    for (final Unit transport : unitToPotentialTransports.get(unit)) {
+      int transportOptions = 0;
+      for (final Unit loadedUnit : TransportTracker.transporting(transport)) {
+        if (containsEquivalentUnit(loadedUnit, unitToPotentialTransports.keySet())) {
+          final Unit equivalentUnit = getEquivalentUnit(loadedUnit, unitToPotentialTransports.keySet());
+          transportOptions += unitToPotentialTransports.get(equivalentUnit).size();
+        } else {
+          transportOptions = Integer.MAX_VALUE;
+          break;
+        }
+      }
+      final double averageTransportOptions =
+          (double) transportOptions / TransportTracker.transporting(transport).size();
+      if (averageTransportOptions < minAverageTransportOptions) {
+        minAverageTransportOptions = averageTransportOptions;
+        selectedTransport = transport;
+      }
+    }
+    return selectedTransport;
+  }
+
+  private static void removeTransportAndLoadedUnits(final Unit transport,
```

Comment:
I think the method is pretty clean. Consider returning the modified map so we can avoid the in/out parameter